WikiWatershed/model-my-watershed
# Allow Projects to Override Layers for Geoprocessing

## Overview

Previously, every layer class (land, soil, etc.) had a single raster layer for it. As of recent work done in #3396 and #3399, we now have allow multiple raster layers for ever class of raster. The actual raster used is specified in the settings, can be overridden, as of #3406.

This PR adds override capability to Projects, in the form of a `layer_overrides` JSON field. The value of this field will be used to override the default layer configuration on every geoprocessing run. For all existing projects, this is set to override the land layer with NLCD 2011. For all new projects, no overrides are set. The front-end is updated to read this value from the project and use it when firing model runs, and the back-end is updated to expect this value in the model runs and use it. Finally, the default land layer is set to NLCD 2019.

This allows us to preserve the work done in all existing projects, thus not changing results from under our users without their awareness, while still using the latest data for all work going forward.

Connects #3400 

## Testing Instructions

Much of the work in this PR is to preserve existing data. Thus, in order to test this, you must create local saved work on `develop`, then checkout this branch and perform the upgrades, then create some new work, and ensure that the old work is preserved.

To compare old and new calculations, we'll use the same Area of Interest in both cases.

* Start on the latest `develop`
* [x] Ensure you locally have a user and are logged in
  - In case registering through the app, follow the Django log to see the confirmation link in the email:
    ```console
    $ vagrant ssh app -c 'sudo journalctl -fu mmw-app'
    ```
* [x] Find an Area of Interest that has noticeable changes between 2011 and 2019
  - I used the CONUS Land Cover Change Index layer in https://www.mrlc.gov/viewer/ to find Flat Rock Creek near Kenbridge, VA that works well. You can search for Kenbridge in the geocoder, go there, then click "Select Boundary", "HUC-12", then "Flat Rock Creek":
    ![2021-07-28 21 33 38](https://user-images.githubusercontent.com/1430060/127418276-9fd8b328-cc42-4ecb-8e77-2cbafcb61511.gif)
* Once analysis is done, go to the Model tab and select "Site Storm Model". Since you are logged in, this project will be saved to the database. Rename the project to be "TR55 Old 2011"
* Under "Details", click "Edit" next to Model to go back one step to Analyze
  ![image](https://user-images.githubusercontent.com/1430060/127419038-d0d115ac-3cb1-4cdd-befc-80837428176b.png)
* Next, select the "Watershed Multi-Year Model". Rename this project to be "MapShed Old 2011"

Now that our reference projects are setup, we can start testing the work in this branch.

* Checkout this branch and run `./scripts/manage.sh migrate` and `./scripts/bundle.sh --debug`
* Clear your geoprocessing cache `vagrant ssh services -c 'redis-cli -n 1 --raw KEYS ":1:geop_*" | xargs redis-cli -n 1 DEL'`
* Restart Celery `vagrant ssh worker -c 'sudo service celeryd restart'`
* Go to http://localhost:8000/ and open the Network tab of developer tools
* Select the same Area of Interest and create a Site Storm Model project named "TR55 New 2019"
  - [x] Notice that the payload for POST http://localhost:8000/mmw/modeling/tr55/ contains a `layer_overrides: {}` key
    ![image](https://user-images.githubusercontent.com/1430060/127419884-70bc90f6-60ad-41d2-871f-29359d6c66b4.png)
* With the same Area of Interest, create a Watershed Multi-Year Model project named "MapShed New 2019"
  - [x] Ensure the payload for POST http://localhost:8000/mmw/modeling/mapshed/ contains a `layer_overrides: {}` key
* Go to your projects (from the top-right), and open the old projects (perhaps in new tabs)
* Compare the output numbers in the tables between the old and new projects
  - [x] Ensure they are different
* In the TR55 Old 2011 project, with the Network tab open, create a new Scenario by clicking "Add changes to this project"
* In the New Scenario, change the Precipitation slider, to re-run the TR-55 model
* In the Network tab, notice that the payload for POST http://localhost:8000/mmw/modeling/tr55/ contains a `layer_overrides: {"__LAND__":"nlcd-2011-30m-epsg5070-512-int8"}` key
  ![image](https://user-images.githubusercontent.com/1430060/127420688-fe7263cf-302a-4d90-9f0e-84687d525850.png)
* Inspect the database with `./scripts/manage.sh dbshell`
  - [x] Ensure that old projects have the 2011 override set, and the new ones don't:
    ```sql
    SELECT id, name, layer_overrides FROM modeling_project;

     id |       name       |                 layer_overrides                 
    ----+------------------+-------------------------------------------------
      1 | TR55 Old 2011    | {"__LAND__": "nlcd-2011-30m-epsg5070-512-int8"}
      2 | MapShed Old 2011 | {"__LAND__": "nlcd-2011-30m-epsg5070-512-int8"}
      9 | TR55 New 2019    | {}
     10 | MapShed New 2019 | {}
    (4 rows)
    ```
